client/src/main/java/org/eclipse/hono/client/impl/GatewayOrDeviceSpecificCommandConsumer.java

```diff
@@ -0,0 +1,344 @@
+/*******************************************************************************
+ * Copyright (c) 2019 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+
+package org.eclipse.hono.client.impl;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+
+import io.vertx.core.AsyncResult;
+import io.vertx.proton.ProtonHelper;
+import org.eclipse.hono.client.Command;
+import org.eclipse.hono.client.CommandContext;
+import org.eclipse.hono.client.HonoConnection;
+import org.eclipse.hono.client.MessageConsumer;
+import org.eclipse.hono.client.ResourceConflictException;
+import org.eclipse.hono.tracing.TracingHelper;
+import org.eclipse.hono.util.CommandConstants;
+import org.eclipse.hono.util.ResourceIdentifier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.opentracing.Span;
+import io.opentracing.SpanContext;
+import io.opentracing.Tracer;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.proton.ProtonQoS;
+import io.vertx.proton.ProtonReceiver;
+
+/**
+ * A wrapper around an AMQP receiver link for consuming commands directed to a gateway or device.
+ * <p>
+ * Includes special support for gateways:
+ * If a gateway creates multiple command subscriptions for specific devices connected to it, the one
+ * <em>GatewayOrDeviceSpecificCommandConsumer</em> instance created for the gateway will contain
+ * multiple {@link CommandConsumerHandler} instances containing the device-specific handlers for
+ * the commands.
+ * <p>
+ * In a non-gateway scenario, the <em>GatewayOrDeviceSpecificCommandConsumer</em> instance will
+ * just contain one {@link CommandConsumerHandler} instance for handling commands to the device.
+ */
+public class GatewayOrDeviceSpecificCommandConsumer extends CommandConsumer {
+
+    private static final Logger LOG = LoggerFactory.getLogger(GatewayOrDeviceSpecificCommandConsumer.class);
+
+    private final Map<String, CommandConsumerHandler> commandConsumerHandlers = new HashMap<>();
+    private final String gatewayOrDeviceId;
+
+    private GatewayOrDeviceSpecificCommandConsumer(final HonoConnection connection, final ProtonReceiver receiver,
+            final String gatewayOrDeviceId) {
+        super(connection, receiver);
```

Comment:
Am I right in assuming that this class doesn't need to be explicitly scoped to a tenant because the receiver link will *always* be scoped to a tenant already?